Dynamodb tutorial

kyleknap opened the pull request:
In this PR, I did the following:
* Created a tutorial directory to add new tutorials.
* Moved the SQS tutorial to the newly created  directory and updated links in that tutorial.
* Made a DynamoDB tutorial. I wrote the tutorial such that you could literally copy and paste each codeblock one after the other in order and successfully get through the entire tutorial.

Let me know if you think.

cc @jamesls @mtdowling

jamesls commented on docs/source/tutorial/dynamodb.rst:
```diff
@@ -0,0 +1,293 @@
+=================
+This tutorial shows how to use
+`Amazon DynamoDB <http://aws.amazon.com/documentation/dynamodb/>`_ with Boto3.
+It will focus on how to use the :py:class:`DynamoDB.ServiceResource` and
```
Instead of focusing on the specific classes, I'd prefer if said the the user is going to accomplish by following this tutorial.  "You'll create a dynamodb table and learn ..."

kyleknap replied:
Sounds good.

jamesls commented on docs/source/tutorial/dynamodb.rst:
```diff
@@ -0,0 +1,293 @@
+    )
+
+    # Wait until the table exists.
+    table.meta.client.get_waiter('table_exists').wait(TableName='users')
```
Didn't realize the waiters weren't mapped in the ddb resource model.  Can we get this updated?  This is an awkward thing to make customers write.

kyleknap replied:
I though that this was awkward too. I will open an issue to track this.

jamesls commented on docs/source/tutorial/dynamodb.rst:
```diff
@@ -0,0 +1,293 @@
+items you want to add, and ``delete_item`` for any items you want to delete::
+
+    with table.batch_writer() as batch:
+        batch.put_item(
```
I think a better example would be to do some sort of for loop that exceeds the batch size, just to drive home the point that we handle the batching for you, e.g. `for i in range(10000)`...

kyleknap replied:
Yeah I could add something like that. I still would want to keep the codeblock for the four items that I have because it is needed for the query and scanning section. I could add a codeblock underneath of it adding a bunch of more users or something. I also link to the method's reference docs and those have an example already in it that uses a large for loop: http://boto3.readthedocs.org/en/latest/reference/services/dynamodb.html#DynamoDB.Table.batch_writer

Let me know what you think. I would be fine adding that codeblock underneath the original example.

jamesls commented:
I would suggest *not* creating a tutorial directory.  This is a lesson learned from boto, where these service specific tutorials never get fleshed out.

Instead I would recommend creating a dynamodb guide that goes under the pre-existing guide/ directory.  We would only have service guides for things that have high level abstractions.

I think eventually we'll make the existing tutorial more generic and less SQS specific, but I would not advise per-service tutorials.

And finally, we should also somehow show the expected output of each call.  We can filter out things like request ids, etc, but I think it's important for the user to see the expected output so they have some indication if they're getting the same results the tutorial gets when they call the various methods.

Content wise, I think it looks good.

kyleknap commented:
@jamesls 

Sounds good about adding output for the tutorial. 

As to removing the tutorial directory, one of the main reasons that I added it there as I think that it would have been weird to see ``DynamoDB Guide`` in the top layer of the scroll bar along with the rest of the guides:
![screen shot 2015-06-25 at 3 45 50 pm](https://cloud.githubusercontent.com/assets/4605355/8367267/622997fc-1b51-11e5-8a9f-47340bcab296.png)

I think I am going to reorganize the listing of guides into a top level ``User Guide`` directory and then mix the DynamoDB specific guide into that directory with some of the other guides.

kyleknap commented:
@jamesls 

I updated the code based on feedback. Here are the changes:
* I grouped the guides into a single section called "User Guides" and categorized the guides.
* I removed the tutorial section and moved the SQS and DynamoDB topics to a "Service Feature Guides" category in the "User Guide" section. I was hesitant in adding the SQS guide back to the featured "tutorial" because I found it awkward that it was the featured tutorial for boto3 when the prose of the entire document is about how to use SQS. It fits much better under a "Service Feature Guides". We will just need to refrain from adding tutorials for services that do not have customizations.
* Added a code snippet about how you can add a bunch of items in using the batch writer. However, that code snippet required a warning to increase write throughput if the code block was attempted to be run because the guide starts with a write throughput of 5.
* Add output to code blocks where applicable.

Let me know what you think.

kyleknap commented:
Also I updated the global toc sidebar to list sections deeper than a depth of 2.

jamesls commented:
+1 to everything except moving the tutorial over to `guide/sqs.rst`.  While I agree with you that the tutorial should get updated, SQS has no high level abstractions so doesn't need a page in the user guide.  I'd rather keep the tutorial in its original location and have one of us update it to be less SQS specific in the near future.

What do you think?

jamesls commented on docs/source/guide/dynamodb.rst:
```diff
@@ -0,0 +1,415 @@
+table.
+
+.. warning::
+    If you are running the code blocks in this guide step by step, it may
```
Rather than adding this note, why don't we just lower the number of iterations.  The max batch size is 25, so even something like 100/500 would demonstrate this.

kyleknap replied:
I would be good with this.

kyleknap commented:
@jamesls 

If I am going to move it back to where it was, I'm going to make some edits for now to make it clear that the SQS tutorial is not the flagship tutorial for boto3 because that is how it looks like right now and I find it really awkward. I will make it clear that is a sample tutorial on how you can use boto3 with an AWS service (not the entire tutorial for boto3).

kyleknap commented:
@jamesls 
I updated the PR. Let me know what you think.

jamesls commented:
:shipit:  Looks good.
